### PR TITLE
[Console] fix parsing escaped chars in StringInput

### LINKED ISSUE #32182: Console\Input\StringInput doesn't handle escaped single quotes correctly
**Symfony version(s) affected**: 2.8-4.3

**Description**  

Symfony\Component\Console\Input\StringInput doesn't work with arguments escaped by `escapeshellarg`.

`escapeshellarg` on a string with single quotes will produce a closing single quote, an escaped single quote and an opening single quote:

```
$arg = 'Jenny\'s';
print($arg . PHP_EOL);
// Jenny's
print(escapeshellarg($arg) . PHP_EOL);
// Jenny'\''s
```

This is correct as per bash documentation: https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Single-Quotes

> A single quote may not occur between single quotes, even when preceded by a backslash.

But this breaks StringInput as it breaks up the string in multiple tokens.

**How to reproduce**  

```
$arg = escapeshellarg('Jenny\'s');
$cmd = 'command --arg=' . $arg;

var_dump($cmd);
var_dump(StringInput::tokenize($cmd));
```

Output:

```
string(26) "command --arg='Jenny'\''s'"
string(26) "command --arg='Jenny'\''s'"
array(4) {
  [0]=>
  string(7) "command"
  [1]=>
  string(11) "--arg=Jenny"
  [2]=>
  string(1) "\"
  [3]=>
  string(1) "s"
}
```

**Possible Solution**  

StringInput should handle escaped characters outside of single quotes part of the string.

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #32182
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Console/Tests/Input/StringInputTest.php [hunk 1]
```diff
@@ -71,6 +71,7 @@ public function getTokenizeData()
             ["--long-option='foo bar''another'", ['--long-option=foo baranother'], '->tokenize() parses long options with a value'],
             ["--long-option='foo bar'\"another\"", ['--long-option=foo baranother'], '->tokenize() parses long options with a value'],
             ['foo -a -ffoo --long bar', ['foo', '-a', '-ffoo', '--long', 'bar'], '->tokenize() parses when several arguments and options'],
+            ["--arg='Jenny'\''s'", ["--arg=Jenny's"], '->tokenize() parses quoted quotes'],
```
Comment (stof) [inline src/Symfony/Component/Console/Tests/Input/StringInputTest.php]: I suggest also adding a case using quoted double quotes, to help avoiding regressions

Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Tests/Input/StringInputTest.php]: like that?

**Review by stof — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by bobthecow** [comment]
@fabpot thanks! will we be getting a v5.3.x release for this as well?

**Comment by fancyweb** [comment]
This patch was released in 5.3.14.

I guess you're talking about https://github.com/symfony/symfony/pull/45240: this one won't be released for 5.3 since 5.3 is no longer maintained, check https://symfony.com/releases/5.3.

**Comment by bobthecow** [comment]
Right, I get that, but this backwards compatibility break _was_ released for 5.3 so we kind of need to fix it.

**Comment by derrabus** [comment]
> so we kind of need to fix it.

We don't. A branch does not have to be bug-free at the time we stop supporting it.

**Comment by bobthecow** [comment]
You're right, "need to" was the wrong choice of words.

A backwards compatibility break was pushed in a 5.3.x release just before the end of the support window. I realized it and sent a fix, also just before the end of the support window. For some reason the fix wasn't released for a month. 

It would be negligent to not revert the BC break in a one off exception to the support policy, or yank the affected releases. I don't particularly care which it is :)

**Comment by fabpot** [comment]
That's an edge case :) The last version was released on January 28. You fixed it on January 29 but I didn't realize it was a BC fix because I wasn't aware of the PR. Anyway, this is the kind of special case where we need to do something. I will see if I can do a 5.3 release.

**Comment by fabpot** [comment]
Just to add more info: The end of maintenance date is "somewhere at the end of the month" and not "the last day of the month". So, other core team members who commented on this issue were right.

But, in the past, whenever we broke BC in the very last release of a minor version, we cherry-picked the fix and made one very last release. That's what I'm going to do here.

(Just wanted to make sure that it's not interpreted as an open door to back-port all the bug fixes)

**Comment by bobthecow** [comment]
That's entirely reasonable. Thanks for the consideration.

**Comment by fabpot** [comment]
Console v5.3.16 released.

**Pull request merged**